File: python/moneysocket/core/message/notification/invoice.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import string
import json

from moneysocket.core.message.notification.notification import (
    MoneysocketNotification)

class NotifyInvoice(MoneysocketNotification):
    MUST_BE_CLEARTEXT = False

    def __init__(self, bolt11, request_reference_uuid):
        super().__init__("NOTIFY_INVOICE",
                         request_reference_uuid=request_reference_uuid)
        self['bolt11'] = bolt11

    @staticmethod
    def cast_class(msg_dict):
        c = NotifyInvoice(msg_dict['bolt11'],
                          msg_dict['request_reference_uuid'])
        c.update(msg_dict)
        return c

    @staticmethod
    def check_valid_msg_dict(msg_dict):
        if 'bolt11' not in msg_dict.keys():
            return "no bolt11 included"
        if type(msg_dict['bolt11']) != str:
            return "unknown bolt11 type"
        #TODO full parse to make sure it includes amount
        if not msg_dict['bolt11'].startswith("lnbc"):
            return "doesn't look like a bolt11"

        return None


MoneysocketNotification.NOTIFICATION_SUBCLASSES['NOTIFY_INVOICE'] = (
    NotifyInvoice)

File: browser/src/js/moneysocket/core/message/notification/invoice.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php


const BinUtl = require('../../../utl/bin.js').BinUtl;
const MoneysocketNotification = require(
    './notification.js').MoneysocketNotification;
let NOTIFICATION_SUBCLASSES = require(
    './notification.js').NOTIFICATION_SUBCLASSES;

class NotifyInvoice extends MoneysocketNotification {
    constructor(bolt11, request_reference_uuid) {
        super("NOTIFY_INVOICE", request_reference_uuid);
        this.bolt11 = bolt11;
    }

    cryptLevel() {
        return "AES";
    }

    static castClass(msg_dict) {
        var c = new NotifyInvoice(msg_dict['bolt11'],
                                  msg_dict['request_reference_id']);
        Object.keys(msg_dict).forEach(key => {
            c[key] = msg_dict[key];
        });
        return c;
    }

    static checkValidMsgDict(msg_dict) {
        if (! ('bolt11' in msg_dict)) {
            return "no bolt11 included";
        }
        if (typeof msg_dict['bolt11'] != 'string') {
            return "unknown bolt11 type";
        }
        // TODO full parse to make sure it includes amount
        if (! msg_dict['bolt11'].startsWith('lnbc')) {
            return "doesn't look like a bolt11";
        }
        return null;
    }
}

NOTIFICATION_SUBCLASSES['NOTIFY_INVOICE'] = NotifyInvoice;

exports.NotifyInvoice = NotifyInvoice;
